### DISCUSSION TITLE
Persistent sid

### DISCUSSION THREAD

**Discussion by se7entyse7en** [miguelgrinberg/python-socketio#1146] [Q&A] [answered]
I was looking for a way to have persistent `sid` and I came up to to [this tutorial](https://socket.io/get-started/private-messaging-part-2/) from the official website.

I tried to understand how would this be possible using the server provided by `python-socketio`, but it seems that it's not possible. If I understand it correctly, the only way would be for the server to have the ability to generate the sid based on the `auth` passed by the client. After digging into the code, I realized that the sid generation is actually the responsibility of the engine [here](https://github.com/miguelgrinberg/python-engineio/blob/main/src/engineio/server.py#L519), though this doesn't have access to the `auth`, so simply subclassing it won't work. Anything I'm missing?

**Comment by se7entyse7en** [comment]
I'm interested in this because I have some data that I fetch during `connect` and assign to the user session using the `sid` as the session key, but I'd like those data to be available also on re-connect, and having a new `sid` won't make it possible. Initially my approach was to just pass the `sessionId` retrieved from the local storage to the `auth` when calling `connect`, and just save the session based on that value instead of the `sid`:

client:
```
const [sessionId] = useLocalStorage("sessionId", "");

useEffect(() => {
    ....

    socket.auth = {
        sessionId
    }
    socket.connect()

    ...
}, [])

```

server:
```
@app.sio.event
async def connect(sid, environ, auth):
    session_id = (auth or {}).get('sessionId')
    if not session_id:
        session_id = str(uuid.uuid4())

    await app.sio.save_session(
        session_id, {'connect_timestamp': int(datetime.now().timestamp())})
```

but it raises `KeyError: 'Session not found'`, so I realized that it cannot be an arbitrary key, but only real `sid`s.

I can workaround it by dealing with a storage manually, but I'm wondering whether there's something that I'm missing.

**Comment by se7entyse7en** [comment]
As a comparison, here's the code of the [TS server of the engine](https://github.com/socketio/engine.io/blob/master/lib/server.ts#L295) where it would be easier to overwrite and apply the logic that I mentioned since it accepts the request.

**Comment by se7entyse7en** [comment]
Though, do I understand correctly from the code that even if a non-in-memory manager such as Redis is used, the user session is still persistend in-memory only?

**Comment by miguelgrinberg** [comment, marked as answer]
Persistent SIDs are not possible. This is true for the JS and the Python versions of the server.

The tutorial you are reading implements a persistent session and user IDs as part of the application. This has nothing to do with the socket `sid` value, which is always unique and does not persist.

> but it seems that it's not possible. If I understand it correctly, the only way would be for the server to have the ability to generate the sid based on the auth passed by the client

Yes, it is possible. Forget about the `sid` value and the user session feature provided by this package, and manage your own sessions, like your tutorial does. I think you are misunderstanding how this tutorial does this. Look at the `sessionStore.js` file, where they show you how to implement session management as part of the application.

When a client connects, it should authenticate. In your tutorial, the client authenticates by passing these made-up Session IDs that reference data stored in the application's session store. In a more production-ready case authentication would be done with a token or API Key. Either way, your server-side `connect` handler should know who the client is using the authentication information. And once it knows who the client is, it can connect this client to the persistent data you stored for that user.

**Reply by se7entyse7en** [reply]
Thanks a lot for the explanation!
